src/build_helper.rs

```diff
@@ -0,0 +1,262 @@
+use core::ops::Range;
+
+use alloc::vec::Vec;
+
+use crate::errors::{DaachorseError, Result};
+
+const INVALID_IDX: u32 = u32::MAX;
+
+/// Helper class in double-array construction to maintain indices of vacant elements and
+/// unused BASE values.
+///
+/// This class manages array elements in fixed-size blocks and supports to extend the array
+/// block by block. Given a constant parameter N, this class maintains information for only
+/// the last N blocks and drops the others according to array extension. We call such last
+/// blocks *active blocks*.
+#[derive(Default)]
+pub struct BuildHelper {
+    items: Vec<ListItem>,
+    block_len: u32,
+    num_elements: u32,
+    head_idx: u32,
+}
+
+impl BuildHelper {
+    /// Creates a helper class that handles the last `block_len * num_free_blocks` elements.
+    pub fn new(block_len: u32, num_free_blocks: u32) -> Self {
```

Comment:
I think block_len should be a constant value because users do not change this value and the constant value can optimize the program.

Reply:
I have a plan to use `BuildHelper` in Charwise as well. In the case of charwise, `block_len` is determined by the alphabet size in runtime, so it is impossible to implement `block_len` as a constant.

If we limit `block_len` to a power of two, we can optimize computations. (But I'd like to do this in another PR)

Reply:
I understand.